### ISSUE TITLE
Should there be support for a finally clause?

### ISSUE THREAD

**Issue by winksaville** [WebAssembly/exception-handling#11]
Some languages such as Java and Python support a finally clause with a definition like that from the [Python spec](https://docs.python.org/3/tutorial/errors.html#defining-clean-up-actions):

> A finally clause is always executed before leaving the try statement, whether an exception has occurred or not.

**Comment by rossberg** [comment]
Finally can be expressed with catch and (re)throw, so isn't required as a primitive. Compilers tend to compile it away (e.g. LLVM, AFAIK).

**Comment by KarlSchimpf** [comment]
I agree with @rossberg-chromium that one can implement the concept, but it requires an "artificial" throw in the try body.

I think a better solution is to define a second

    try inst* finally inst* end

block that always evaluates the finally instructions. This cleanly separates the two types of semantics, keeping things simple.

**Comment by winksaville** [comment]
I might be totally wrong here, but it seems to me that `wasm` is an intermediate representation of compiler output. As such, it will be used to combine separately compiled entities from different compilers. Therefore, it would seem better provide to the notion of `finally` as a primitive so the semantics of the `finally` clause can be unambiguously communicated.

**Comment by rossberg** [comment]
@winksaville, not quite, Wasm is an abstraction defining a virtual ISA. Automatic interoperability between different compilers is explicitly a non-goal (they would need to collaboratively define a common ABI).

**Comment by winksaville** [comment]
@rossberg-chromium, it seems to me that `finally` has very definite semantics that is intertwined with the  try/catch/else implementation and semantics. So defining it as @KarlSchimpf proposes [above](https://github.com/WebAssembly/exception-handling/issues/11#issuecomment-300884843) or some other way would be beneficial even within a system that's built by a single compilation.

**Comment by KarlSchimpf** [comment]
Would it be ok to consider a `try inst* finally' inst* end` (as I suggested above) to be a _future_ feature? Or, do you  feel that this must be baked in this initial, minimal proposal?

**Comment by winksaville** [comment]
IMHO it should be in the MVP as `try finally` is integral to several languages, but it's your call.

Question, are you sure `try/finally` can be implemented in terms of `try/catch` since `wasm` is limited to [structured control flow](https://github.com/qwertie/wasm-design/blob/master/Rationale.md#control-flow)?

Also, reading [exception handing](https://en.wikipedia.org/wiki/Exception_handling) on wikipedia there seems to be at least a implementation schemes, dynamic registration and table driven. Have you implemented anything yet?

**Comment by KarlSchimpf** [comment]
First off, "structured control flow" is relatively easy to add. This is how
LLVM (and emscripten that uses LLVM) convert C/C++ CFG code into structured
control flow.

It involves two steps in LLVM. The first converts an irreducible graph to
an reducible one by introducing a local variable to hold the label to jump
to, and an "artificial" loop construct for the control flow. This rarely
has to be done, but is there for the few cases it is needed.

Then it runs what is called a "relooper" that extracts out the blocks/loops
(using dominator graphs of the CFG). These blocks are build (nested) in
such a way that all branches can be converted to a WASM branch. The
br_table instruction implements table branches (which can be used for both
switches and the tables for exception handling).

Once this is done, WASM structured code is straight forward. It should also
be noted that you can convert WASM structured code back into the original
CFG.

This design was intentional. It allows most Javascript jit-compilers, which
require structured code to optimize, to be simply ported to WASM. This was
the "great" compromise to this problem. However, we have been careful to
allow the capability of converting WASM structured code back to a CFG, so
that more traditional compiler techniques can be used.

As far as handling exceptions (with the notion of try-finally blocks) our
group has internally implemented a prototyped version in LLVM. This version
takes the standard (table-driven) exception handling and converts it to
structured code using the relooper.

Note: Without try-finally blocks, there is more duplication of code, but
that is the only price paid.

The key trick is how you define the thrown exception. Rather than actually
throwing the exception, only the 'index' into the exception table handler
is thrown, and then the catch code does the rest of the interesting
dispatching. Hence, we have shown that this concept can be implemented in a
traditional compiler using traditional techniques.

Note that means for the LLVM C compiler, it will have at most two catch
blocks. The first is a tagged (imported) exception to capture C++
exceptions. The other is a default catch block to handle exceptions from
another language.

Hence, we have full confidence that this proposal will work. We also
recently hired a new employee to move our prototype implementation into
LLVM for the WASM target.

Karl Schimpf

**Comment by winksaville** [comment]
SG and I think your suggestion of separating `try finally` from `try catch` is a good one.

**Comment by KarlSchimpf** [comment]
Marked as (future) enhancement to this proposal.

**Comment by Abion47** [comment]
Out of curiosity, I looked at the IL generated from compiling the following C# script:

```
try {
  ...
} catch {
  ...
} finally {
  ...
}
```

The IL byte code that the compiler generated looked like this (ish):

```
try {
  try {
    ...
  } catch {
    ...
  }
} finally {
  ...
}
```

So it seems like there may be something to this idea of separating the `try catch` and the `try finally` blocks.

**Comment by bartwe** [comment]
One additional issue with finally is that it gets called also when within the try block a 'return' call is made.

**Comment by rossberg** [comment]
@bartwe, in Wasm, `return` is just a special case of a `br`, where the same applies. The same semantics exists in e.g. C++, yet, compilers like clang on LLVM compile away `finally`.

**Comment by osa1** [comment]
Should this be closed in favor of #158?

**Comment by aheejin** [comment]
Yeah I think this can be closed. I don't think we will have `finally` in the current version of the EH proposal anyway though.
